hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/operations/AbstractRegistrationOperation.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.impl.eventservice.impl.operations;
+
+import com.hazelcast.internal.cluster.ClusterService;
+import com.hazelcast.internal.cluster.Versions;
+import com.hazelcast.internal.cluster.impl.ClusterTopologyChangedException;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.nio.serialization.impl.Versioned;
+import com.hazelcast.spi.ExceptionAction;
+import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.impl.AllowedDuringPassiveState;
+import com.hazelcast.spi.impl.SpiDataSerializerHook;
+
+import java.io.IOException;
+
+import static java.lang.String.format;
+
+abstract class AbstractRegistrationOperation extends Operation
+        implements AllowedDuringPassiveState, IdentifiedDataSerializable, Versioned {
+
+    private int memberListVersion = -1;
+
+    AbstractRegistrationOperation() {
+    }
+
+    AbstractRegistrationOperation(int memberListVersion) {
+        this.memberListVersion = memberListVersion;
+    }
+
+    @Override
+    public final void run() throws Exception {
+        runInternal();
+        checkMemberListVersion();
+    }
+
+    protected abstract void runInternal() throws Exception;
+
+    private void checkMemberListVersion() {
+        ClusterService clusterService = getNodeEngine().getClusterService();
+        if (clusterService.isMaster()) {
```

Comment:
Shouldn't we also test for `memberListVersion > -1` here? On a 3.9 member working in 3.8 cluster version, `readData` will not deserialize any member list version.

Reply:
right. I'll fix that.